File: fsharp/proverb/Proverb.fs
module Proverb
let recite (input: string list) = 
    match input with
    | [] -> []
    | [one] -> [$"And all for the want of a {one}."]
    | _ ->
        let shiftedInput = input |> List.permute (fun x -> (x + List.length input - 1) % List.length input)
        let shiftedOutput = shiftedInput |> List.map2 (fun x y -> $"For want of a {x} the {y} was lost.") input |> List.take (List.length input - 1)
        [$"And all for the want of a {List.head input}."] |> List.append shiftedOutput

File: dart/proverb/lib/proverb.dart
class Proverb {
  String recite(List<String> items) => items.indexed
      .map((entry) => entry.$1 < items.length - 1
          ? 'For want of a ${entry.$2} the ${items[entry.$1 + 1]} was lost.'
          : 'And all for the want of a ${items[0]}.')
      .join('\n');
}
